src/safeposix/syscalls/net_calls.rs

```diff
@@ -755,73 +749,75 @@ impl Cage {
-                            let mut sockobjthere = sockobjtherelock.write();
-
-                            //dropping the old socket closes it
-                            sockobjthere.0 = interface::Socket::new(sockfdobj.domain, sockfdobj.socktype, sockfdobj.protocol);
-                            sockobjthere.1 = ConnState::NOTCONNECTED;
-
-                            if let Some(localaddr) = sockfdobj.localaddr {
-                                if sockobjthere.0.bind(&localaddr) != 0 {
-                                    panic!("Bind on known ok address failed within shutdown!");
-                                }
-                            }
+            if let Some(soid) = sockfdobj.socketobjectid {
+                if shutdown {
 
-                            //check reuseaddr/port
-                            for optname in [sockfdobj.options & SO_REUSEPORT, sockfdobj.options & SO_REUSEADDR] {
-                                if optname != 0 {
-                                    if sockobjthere.0.setsockopt(SOL_SOCKET, optname, 1) < 0 {
-                                        panic!("Setsockopt within known ok conditions failed within shutdown!");
-                                    }
-                                }
-                            }
+                    let sockobjtherelock = NET_METADATA.socket_object_table.get(&soid).unwrap().clone();
+                    let mut sockobjthere = sockobjtherelock.write();
+                    let shutresult = sockobjthere.0.shutdown(how);
 
-                            //check nonblock
-                            if sockfdobj.flags & O_NONBLOCK != 0 {
-                                if sockobjthere.0.set_nonblocking() < 0 {
-                                    panic!("Setting nonblock using fcntl on known ok fd failed within shutdown!");
-                                }
-                            }
+                    if shutresult < 0 {
+                        match Errno::from_discriminant(interface::get_errno()) {
+                            Ok(i) => {return syscall_error(i, "shutdown", "The libc call to setsockopt failed!");},
+                            Err(()) => panic!("Unknown errno value from setsockopt returned!"),
+                        };
+                    }
 
-                            //now we have completely recreated the socket but unconnected
-                        } else {
-                            //Reaching this means that the socket is closed. Removing the sockobj
-                            //indicates that the sockobj will drop, and therefore close
-                            NET_METADATA.socket_object_table.remove(&soid).unwrap();
+                    match how {
+                        SHUT_RD => {
+                            if sockobjthere.1 == ConnState::CONNRDONLY { releaseflag = true; }
+                            sockobjthere.1 = ConnState::CONNWRONLY;
```

Comment:
this feels like it should be in an else. I don't know, the code makes sense to me, but I would hate to have an issue where we change the connection to a wronly and the releaseflag is for some reason not set and then the connection remains open as a wronly